karapace/schema_registry_apis.py

```diff
@@ -395,16 +420,39 @@ async def subjects_list(self, content_type):
         subjects_list = [key for key, val in self.ksr.subjects.items() if self.ksr.get_schemas(key)]
         self.r(subjects_list, content_type, status=HTTPStatus.OK)
 
-    async def subject_delete(self, content_type, *, subject):
-        self._subject_get(subject, content_type)
-        version_list = list(self.ksr.get_schemas(subject))
+    async def subject_delete(self, content_type, *, subject, request: HTTPRequest):
+        permanent = request.query.get("permanent", "false").lower() == "true"
+
+        subject_data = self._subject_get(subject, content_type, include_deleted=permanent)
+
+        if permanent is True and [
+            version for version, value in subject_data["schemas"].items() if value.get("deleted", False) is False
+        ]:
+            self.r(
+                body={
+                    "error_code": SchemaErrorCodes.SUBJECT_NOT_SOFT_DELETED.value,
+                    "message": f"Subject '{subject}' was not deleted first before being permanently deleted",
+                },
+                content_type=content_type,
+                status=HTTPStatus.NOT_FOUND,
+            )
+
+        version_list = list(self.ksr.get_schemas(subject, include_deleted=permanent))
         if version_list:
             latest_schema_id = version_list[-1]
         else:
             latest_schema_id = 0
         are_we_master, master_url = await self.get_master()
         if are_we_master:
```

Comment:
The flag for permanent delete is not forwarded if we are *not* the master. (couldn't comment on the line because GH UI doesn't allow it)